packages/vite/src/node/server/middlewares/indexHtml.ts

```diff
@@ -86,25 +87,24 @@ export function createDevHtmlTransformFn(
     html: string,
     originalUrl?: string,
   ): Promise<string> => {
+    const normalized = normalizeIndexHtmlUrl(url)
+    let filename: string
+    if (normalized.pathname.startsWith(FS_PREFIX)) {
+      filename = fsPathFromId(normalized.pathname)
+    } else {
+      filename = normalizePath(
+        path.join(server.config.root, normalized.pathname),
+      )
+    }
     return applyHtmlTransforms(html, transformHooks, {
-      path: url,
-      filename: getHtmlFilename(url, server),
+      path: normalized.url,
```

Comment:
I think we shouldn't normalize `path` here. We don't know if `/` is responding with the same HTML file with `/index.html` when `appType: 'custom'`.
Instead, I think we should make `devHtmlHook` function work with `path` ending with `/`.
https://github.com/vitejs/vite/blob/3734f8099e3922c189497ce404fe7ff2f8929ae1/packages/vite/src/node/server/middlewares/indexHtml.ts#L185

Reply:
My reasoning was that the path should be the same during build and development. And I assumed that the paths always refer to a file during build. 

Can the html path be `/` during build?

Reply:
> Can the html path be `/` during build?

No, it always ends with `.html` during build currently. But it's valid to only call `transformIndexHtml` in dev and not use it in build.

Reply:
If that's the case, if no constraints are placed on the path, shouldn't the user decide filename as well?

Reply:
> In that case, if no constraints are placed on the path, shouldn't the user decide filename as well?

Yes, I was thinking of adding `file` parameter to `server.transformIndexHtml`. Something like:
```ts
type TransformIndexHtmlInput = {
  /** the served URL path (may include queries and hashes) (should include base if exists) */
  url: string
  /**
   * the file path to the HTML file
   * used to resolve the relative URLs for script tags and other assets
   */
  file: string
}

interface ViteDevServer {
  transformIndexHtml(input: TransformIndexHtmlInput): Promise<string>
  /** @deprecated -- kept for backward compat */
  transformIndexHtml(
    url: string,
    html: string,
    originalUrl?: string,
  ): Promise<string>
}
```
But this requires more code changes. It would be awesome if you want to work on this (no pressure, I'm also fine only with the fix).